src/Formats/MarkInCompressedFile.h

```diff
@@ -40,20 +34,84 @@ struct MarkInCompressedFile
 
     String toStringWithRows(size_t rows_num) const
     {
-        return "(" + DB::toString(offset_in_compressed_file) + "," + DB::toString(offset_in_decompressed_block) + "," + DB::toString(rows_num) + ")";
+        return "(" + DB::toString(offset_in_compressed_file) + "," + DB::toString(offset_in_decompressed_block) + ","
+            + DB::toString(rows_num) + ")";
     }
-
 };
 
-class MarksInCompressedFile : public PODArray<MarkInCompressedFile>
+/**
+ * In-memory representation of an array of marks.
+ *
+ * Uses an ad-hoc compression scheme that decreases memory usage while allowing
+ * random access in O(1) time.
+ * This is independent from the marks *file* format, which may be uncompressed
+ * or use a different compression method.
+ */
+class MarksInCompressedFile
 {
 public:
-    explicit MarksInCompressedFile(size_t n) : PODArray(n) {}
+    using PlainArray = PODArray<MarkInCompressedFile>;
+
+    // Takes the sequence {marks[base], marks[base + stride], marks[base + stride * 2], ...}
+    // and converts it to the compressed in-memory representation.
+    MarksInCompressedFile(const PlainArray & marks);
+
+    MarkInCompressedFile get(size_t idx) const;
 
-    void read(ReadBuffer & buffer, size_t from, size_t count)
+    size_t approximateMemoryUsage() const;
+
+private:
+    /** Throughout this class:
+     *   * "x" stands for offset_in_compressed_file,
+     *   * "y" stands for offset_in_decompressed_block.
+     */
```

Comment:
(This renaming seems goofy, but I think it's good. It was pretty annoying to read otherwise. The main downside is that the reader may miss this comment and be confused. I'm ok with reverting to full names if you feel it's more confusing this way.)

Reply:
yeah, this is spicy :)